cqc/pythonLib.py

```diff
@@ -380,7 +623,19 @@ def close(self, release_qubits=True):
             self.flush()
 
         if release_qubits:
-            self.release_all_qubits()
+            if len(self.active_qubits[:]) != 0:
```

Comment:
Why not call `release_all_qubits`?

Reply:
The sequence with `release_all_qubits` went like this:

```
def close(self, release_qubits=True):
        """Handle closing actions.
        Flushes remaining headers, releases all qubits, closes the 
        connections, and removes the app ID from the used app IDs.
        """

        # Flush all remaining commands
        if self._pending_headers:
            self.flush()

        if release_qubits:
            self.release_all_qubits()
        self._s.close()
        self._pop_app_id()

        self.closeClassicalServer()

        for name in list(self._classicalConn):
            self.closeClassicalChannel(name)
```

```
def release_all_qubits(self):
        """
        Releases all qubits off this connection
        """
        return self.release_qubits(self.active_qubits[:])
```

```
def release_qubits(self, qubits, notify=True, block=True, action=False):
        """
        Release qubits so backend can free them up for other uses
        :param qubits: a list of qubits to be released
        :param notify:     Do we wish to be notified when done.
        :param block:         Do we want the qubit to be blocked
        :param action:     Execute the releases recursively or sequencely
        """

        if isinstance(qubits, qubit):
            qubits = [qubits]
        assert isinstance(qubits, list)
        n = len(qubits)

        if n == 0:  # then we don't need to do anything
            return

        logging.debug("App {} tells CQC: Release {} qubits".format(self.name, n))
        if action:
            hdr_length = CQCCmdHeader.HDR_LENGTH + CQCSequenceHeader.HDR_LENGTH
        else:
            hdr_length = CQCCmdHeader.HDR_LENGTH
        hdr = CQCHeader()
        hdr.setVals(CQC_VERSION, CQC_TP_COMMAND, self._appID, hdr_length * n)
        cqc_msg = hdr.pack()

        release_messages = b""

        for i in range(n):
            q = qubits[i]
            try:
                q.check_active()
            except QubitNotActiveError as e:
                raise QubitNotActiveError(
                    str(e) + ". Qubit {} is not active. None of the qubits are released".format(q._qID)
                )
            q._set_active(False)
            cmd_hdr = CQCCmdHeader()
            cmd_hdr.setVals(q._qID, CQC_CMD_RELEASE, int(notify), int(block), int(action))
            release_messages += cmd_hdr.pack()
            if action:
                seq_hdr = CQCSequenceHeader()
                # After this one we are sending n-i-1 more releases
                seq_hdr.setVals(hdr_length * (n - i - 1))
                release_messages += seq_hdr.pack()

        self._s.send(cqc_msg + release_messages)

        if notify:
            msg = self.readMessage()
            self.check_error(msg[0])
            if msg[0].tp != CQC_TP_DONE:
                raise CQCUnsuppError(
                    "Unexpected message sent back from the server. Message: {}".format(msg[0].printable())
                )
            self.print_CQC_msg(msg)
```

So `release_all_qubits()` isn't really needed, since you might as well just do `release_qubits(self.active_qubits[:])`

`release_qubits()` has been replaced by `construct_release()`, `commit()`, and the `if notify:` block. This fits better in the model of first constructing a message and then sending it.

Reply:
Ok!